### ISSUE TITLE
Download HEFER page

### ISSUE THREAD

**Issue by CrispinF** [exeGesIS-SDM/ELM#19]
Similar to the page on the CS portal.
However if there are notifications, the page perhaps needs to report on SBI not found + SBI HEFER in preparation, and HEFER ready for download.
Should this still be available to unauthenticated users? Worth checking this with RPA.

**Comment by sarahpoppy** [comment]
No reason to change from current behaviour.  This should be available to non registered users entering a SBI, and give two returns, i.e. download the HEFER or HEFER not found.  The in preparation is only a matter of hours surely, so I don't think in prep is needed.

**Comment by CrispinF** [comment]
I think it would lead to confusion to say "not found" when it is in prep. An RPA person might test this and think the system was not working if they had entered the SBI earlier. It's no work to make it say that an item is in preparation.

**Comment by sarahpoppy** [comment]
okay

**Comment by CrispinF** [comment]
Note that on the existing https://www.myshinedata.org.uk/retrieve-hefer an admin (such as Sarah) can edit the text above the download controls.
@sarahpoppy how important is that capability? For example, would you be happy to specify that text up front and any change would require a code deployment? That's one extreme. At the other, you'd be able to edit the text in situ at any time. The latter is obviously a bit more work than the former, but probably less over time.

**Comment by sarahpoppy** [comment]
I think the functionality for me to edit is important so worth the investment upfront.  its likely this will want to change over time

**Comment by CrispinF** [comment]
@JimKerslake I think this is looking like a custom page with a text pack and _in situ_ text editing capability.
The texts might break down into 3 items:

1. Page heading (single line plain text)
2. Intro text (html)
3. Control label (single line plain text)
4. Not found message ((html) - not editable in situ as it only appears for a not found item.

?? @sarahpoppy is there visible feedback when an SBI is found? Or just the actual download?

**Comment by JimKerslake** [comment]
New page added at  https://elm-hefer-staging.esdm.co.uk/downloadhefer 
(and added a menu item for it via the CS page management admin system)

Any user who fulfils the auth policy 'ConfigurationDisplayEditingLinksPolicy'
will see little 'TM' links beside (or beneath) the 3 editable text resources on this page, which link to the text/html editors for those resources.

(In retrospect I have no idea why we chose the letters 'TM' for these links - 
I may go back and re-work the inline editor component to say something more sensible like 'edit'.)

Any user who fulfils the auth policies 'ConfigurationTextEditorPolicy' and 'ConfigurationManagementPolicy'
can alternatively go via the admin > configuration system to edit all 4 strings (that is the 'nothing found' string too) 
via the configuration provider.

**Comment by CrispinF** [comment]
Note that only the latest instance for an SBI should be presented for download (see issue #8 )

**Comment by CrispinF** [comment]
Note that we must do the lookup by SBI and user. Sarah and Bob might request the same SBI at different times, and we need to check for the right one.

**Comment by CrispinF** [comment]
@sarahpoppy input on the name of the packs needed.
At present we've coded this into the URL "downloadhefer" which would not be desirable if "HEFER" is not the right name.
If there is still no name, maybe we should change this to something more generic, like "pack-download"?

(other texts will be editable in situ).

**Comment by sarahpoppy** [comment]
I would make the coding terminology generic ideally, and we can issue the naming cosmetically, but at the same time not if this is lots of additional work to undo work already done.   

The RPA have settled on the name of SFI HEFER in the absence of considering any alternatives as of yesterday, despite numerous challenges from ourselves, and it just remains to be whether we can be bothered to challenge yet again.   Will be resolved finally tomorrow.

**Comment by CrispinF** [comment]
@joerichardsESDM retain existing route /downloadhefer
Texts on the page are editable.

**Comment by CrispinF** [comment]
@joerichardsESDM could you change menu item text to "Download".

**Comment by CrispinF** [comment]
@joerichardsESDM as discussed earlier, would be good if this page can take SBI number in route/querystring to auto-populate the SBI control and run the request. This is not a "must have" though, obviously the retrieving of the right record and the download is more important.

Also, regarding permissions, ideally a SBI download should be accessible either because you requested it, or because you are in the "Application Administrators" or "Administrators" roles (i.e. those roles can access any SBI). Often we'd do such things with Authorisation Policies, but this is so simple a role check is sufficient I think.

Achieve these things will stop us having to add download capabilities anywhere else.

**Comment by joerichardsESDM** [comment]
Will be working on staging site once code is merged & CI runs, including ability to add SBI number as a route parameter which skips the input form. I'll have to revisit permissions as I hadn't read Crispin's last until now.

**Comment by joerichardsESDM** [comment]
Valid SBIs for testing purposes (these are taken from myshinedata.co.uk):
106230844
106974800
106819618
107197195
106189581	
106788121
107193395
106300565	
107332064	
106894040

**Comment by CrispinF** [comment]
@joerichardsESDM maybe it's not supposed to be working yet, but my first test went in, then the download page said it was awaiting land parcels, and then a little later it said "No HEFER request with SBI requested by Crispin-gmail can be found in the database." (which is obviously incorrect or something has gone wrong).
The SBI was 106230844

**Comment by CrispinF** [comment]
@sarahpoppy note we now have an end-to-end working process, on staging. The pack has nothing meaningful in it yet, but that comes next.

For me the download page needs work so the user can check again without having to reload the page and re-enter their SBI - that's a complete pain.  I think the SBI input should stay on the screen after the check, with the button to execute the check, if the result is anything other than a download button.

**Comment by CrispinF** [comment]
@sarahpoppy and now some base maps are appearing in the PDF for download. An example SBI is 107000029

And this pack can now be made available as a link like this in an email https://elm-hefer-staging.esdm.co.uk/download-hefer/107000029
which requires the user to sign in, and then presents the download button.
